Add time zones support

straight-shoota opened the pull request:
Until now, instances of `Time` had a `kind` member which indicated, whether this instance holds a time in UTC, an unspecified "local" timezone or is entirely unspecified.

This PR adds a `Time::Location` class which represents a specific time zone. Each `Time` instance references a `Location` (i.e where this time applies). From this location's transition rules, the applicable offset at each point in time can be derived, including daylight savings time, time zone changes etc.

Storing only an offset from UTC is not enough, because from that you can't infer what timezone transition rules to apply. And offset changes inside a time zone happen quite frequently (for example daylight savings).

# Time
The `kind` ivar in `Time` is replaced by `location` which is a pointer to a `Location` instance.

```crystal
struct Time
  @seconds : Int64                    # 8 Byte
  @nanoseconds : Int32                # 4 Byte
  @kind : Time::Kind                  # 4 Byte
end

# Becomes:
struct Time
  @seconds : Int64                    # 8 Byte
  @nanoseconds : Int32                # 4 Byte
  @location : Pointer(Time::Location) # 8 Byte (64 bit architecture)
end
```

This means, each `Time` struct will require 24 bytes instead of the currently used 16 bytes (on 64-bit). This increases the size of a time struct slightly but it is inevitable to embed a reference to a time zone if proper time zone handling should be supported.

For incremental time (a reference to a a specific point in time), `seconds` and `nanoseconds` always represent time in UTC. The offset will only be added when needed (for example to access one of the date/time fields individually). This makes it easier than if it was stored *with* offset as that might lead to invalid or ambiguous instances (for example when an hour is skipped or happens twice at a DST switch). Another benefit is, that retrieving the offset (zone transitions are looked up by unix epoch) and comparing two instances requires no conversion. Converting a time between zones means just exchanging the `location`.

# Time::Location

Each location can be one of the following
* a timezone in the IANA time zone database read from a local copy of this database which is provided for by the operating system (available in most unix-like OS's) or a custom distribution (can be configured throught `ZONEINFO` environment variable)
* A zone with a fixed-offset
* Coordinated Universal Time, available as `Time::Location::UTC`
* an unspecified time zone, indicating a floating time which is not tied to a specific time zone.

The concept for loading location data from system sources is inspired by [Go's implementation](https://golang.org/pkg/time/#Location) which seems very reasonable for this.

Of course, it can't be expected to find that zoneinfo data available on every platform (hey, Windows!). Other systems will probably require some additional effort, but Go has implementations for a variety of platforms, so it can be done one way or another. As far as I understand, under Windows, timezone information is available (needs some translation and stuff, but it's usable) but only limited to the current applicable time zone data. Application that need to cover wider time frames will probably have to distribute their own copy of the timezone database. But in general, it is certainly best if we can rely on the operating system to provide that data.
For now, it should work for all currently supported Crystal targets.

# Floating time

A floating time which is not tied to a specific time zone is currently indicated by the special location `Location::UNSPECIFIED`. Time instances created with `Time.new` and no location argument are by default floating times. This means, they don't represent a specific point in time but are only aplicable in a specific context.
A floating time `08:30` can for example represent each point in time, where the local time shows `08:30`, independent of local time's timezone. If I set an alarm at `08:30`, I want that alarm to trigger when it is `8:30` wherever I am at that point.
But it doesn't have to be that, it can also have a different meaning. Floating times have no semantic context.

Comparing or operating on floating and incremental times causes problems:

For example, my computer time says `2017-11-24 19:37`, we'll use that as a floating time.
Another source tells me, the current time in UTC is `2017-11-24 18:37 UTC` or in my local timezone: `2017-11-24 19:37 +01:00`.
Let's compare that with the floating time:

```crystal
floating = 2017-11-24 19:37
utc      = 2017-11-24 18:37 UTC
local    = 2017-11-24 19:37 +01:00

utc      == local # => true, both represent the same point in time
floating == utc   # => ?
floating == local # => ?
```

We know that all three readings have been observed a the same point in time, so the last two comparisons should both be true, but they are mutually exclusive.

These issues with floating time are present all over the place (what should the format `%z` (time zone) print?). I've tried to implement this as sane as possible, but it's still a conceptual problem.
To solve this, I've already brought up the proposal to add a dedicated type for floating time, but this is another issue.

# Time representation
This PR also modifies the string representation of a `Time` instance (regarding `#inspect`, `#to_s`):

```yaml
#  Floating times with unspecified location have no time zone attached and won't print any.
2017-11-25 22:06:17

# UTC times print time zone `UTC`
2017-11-25 22:06:17 UTC

# Times with a fixed-offset location print the timezone offset (seconds are truncated unless relevant)
2017-11-25 22:06:17 +01:00
2017-11-25 22:06:17 +01:00:37

# Times with a non-fixed-offset location print the timezone offset plus zone name
2017-11-25 22:06:17 +01:00 Europe/Berlin
```

Requires #5321
Fixes #2490
/cc #2481 #4556

RX14 commented:
I don't think that times should have an unspecified timezone by default, in fact I think we should reconsider the existence of an unspecified time zone: does it have a valid usecase. Having the possibility of not being able to convert a given time to an instant in UTC adds a lot of error cases.

straight-shoota commented:
`Time.new` could probably use the local timezone by default, but I'm not entirely sure about this. For now I've kept it as it has been before (only that until now it didn't have much effect anyway).
It could also make sense to promote the more descriptive constructor methods supporting date field arguments - namely `Time.utc`, `Time.local` (maybe `Time.floating` or similar) - and maybe make `Time.now` to require `location` instead of providing a default. This would be super expressive but I guess it's not as developer-friendly and easy-to-understand for beginners.

I'd favor to keep this PR focused on the implementation of time zones and have the general discussion about floating time in a separate issue. For now, I've tried to adjust the existing structure to the new concept, but this can be changed depending on the outcome of the floating time issue.

straight-shoota commented:
I've added a section about *Time representation* in the OP

Sija commented on src/time.cr:
```diff
@@ -510,8 +510,12 @@ struct Time
+      else
+        Format.new(" %::z").format(self, io)
+      end
+      io << " " << location.name unless location.fixed?
```
`" "` could be a `Char`.

straight-shoota replied:
Thanks, I intended to change that.
However, I'd rather welcome feedback on the general concept instead of such minor details ;)

luislavena commented:
Thank you @straight-shoota for this PR, your approach looks very interesting and will love to take a look and see this implementation against some apps I have in the following days.

Will be possible for you to include some examples on creation of time instances on fixed locations? For example, how will be possible to recreate certain date/time on EST/EDT or PST/PDT (perhaps adding that to the `Time.new` constructor documentation).

Also notice that `Time.new` with no argument will default to `Time::Location.local` while `Time.new` with arguments will default to unspecified (floating). It will be great to include more examples on that usage here in the PR or the code itself (I prefer the later).

Once again, thank you for your contribution and looking forward your responses!
❤️ ❤️ ❤️

straight-shoota commented:
@luislavena Some typical usage should be documented in the specs - but of course the documentation still needs to be improved further.

Some examples:
```crystal
# using a fixed offset of 3:00:
location = Time::Location.fixed(3 * 3600)
time =  Time.new 2014, 10, 30, 21, 18, 13, location: location # => 2014-10-30 21:18:13 +03:00
# using a tz database location:
location = Time::Location.load("Europe/Berlin")
time = Time.new 2014, 10, 30, 21, 18, 13, location: location # => 2014-10-30 21:18:13 +01:00 Europe/Berlin
location = Time::Location.load("EST5EDT")
time = Time.new 2014, 10, 30, 21, 18, 13, location: location # => 2014-10-30 21:18:13 -04:00 EST5EDT
location = Time::Location.load("PST8PDT")
time = Time.new 2014, 10, 30, 21, 18, 13, location: location # => 2014-10-30 21:18:13 -07:00 PST8PDT
```

The use of `Time::Location.load/Time::Location.fixed` might be a bit tedious, but I'm not sure if and how this could be improved. It would be possible to allow `location` to receive a string or int and create the corresponding location accordingly. But I'm not sure if this would be a good idea as it would hide the fact that each call creates a new instance of `Time::Location`.

`Time.new` is essentially the same as `Time.now`, and for the current time it makes sense to be in the local timezone by default. `Time.new` with individual date/time fields is something different, and as long as there is no location provided, it makes sense to recognize this as "floating" time.  It usually makes little sense to create a reference to an absolute wall time reading with a completely unknown zone offset - for all we know, `Location.local` could be *anything* at runtime, so it is essentially a floating date, not a specific instance in incremental time.
If you want to create a reference to an absolute point in time, you need to specific an explicit time zone reference, either by using `Time.utc` or specifying a concrete location (either fixed offset or named location).

So, clearly, the naming can be a bit confusing and should certainly be improved, but that will probably depend on what we decide to do about floating time. One solution could be to make `Time.new` always require a `location` and introduce a different means to create floating time (like `Time.floating`) which won't accept a different location.

bew commented:
Is it possible/want-able to create a `Location` object from a `Time::Span` object ?

straight-shoota commented:
@bew Do you have some usecase in mind?
It should be as easy as `Location.fixed(span.total_seconds)`, so I don't think it would be worth adding a specific API for this.

bew commented:
No, no use-case in mind, I was just reading your example `Time::Location.fixed(3 * 3600)` and was thinking "hmm could we do `Time::Location.fixed(3.hours)` instead?"

straight-shoota commented:
Yeah, I think it should not be too common to use this constructor at all, and even less with a static value.
Creating fixed-offset locations should be avoided as much as possible anyway, unless there is no other way, for example when the time is read from an external source with only an offset and no time zone location attached.

Such absolute timepoints in machine-readable data don't need any offset at all and should better work directly in UTC.
Offsets are only relevant for human interaction, but then you need a proper timezone location anyway to accurately map time and offsets.

luislavena commented:
@straight-shoota will be possible for you to rebase against master now that #5321 has been merged?

Thank you.
❤️ ❤️ ❤️

straight-shoota commented:
Rebased.

luislavena commented:
Thank you @straight-shoota for the explanation and sorry for the delay on getting back to you on your comments.

> re: floating time

From the example:

```crystal
floating = 2017-11-24 19:37
utc      = 2017-11-24 18:37 UTC
local    = 2017-11-24 19:37 +01:00

utc      == local # => true, both represent the same point in time
floating == utc   # => ?
floating == local # => ?
```

I don't think `floating == utc` should return `true`. As described, the scenario of least surprise will be that floating compares the time and date elements of both dates disregard of the time location of each other. 

That will allow me ask the question _tell me when is 7pm_ independently of the time location the _other_ time instance I'm comparing against it.

>But I'm not sure if this would be a good idea as it would hide the fact that each call creates a new instance of `Time::Location`.

This will definitely need to be addressed via a cache per-process. Looked at this through the eyes of someone that had to load 6 different timezones and process 500K date records. Haven't converted the script to test this (is a Ruby one) but might do over the weekend for further comments.

Last but not least, really love the flexibility to move time instances across timezones:

```crystal
ny = Time::Location.load("America/New_York")
ar = Time::Location.load("America/Buenos_Aires")

time1 = Time.now(location: ny)
pp time1 # => 2017-11-28 10:34:23 -05:00 America/New_York

time2 = time1.in(location: ar)
pp time2 # => 2017-11-28 12:34:23 -03:00 America/Buenos_Aires
```

❤️ ❤️ ❤️

straight-shoota commented:
Well, `utc` and `local` describe the same point in time, so why `floating` should really be equal `utc` when both are observed in the same time zone. This can't be decided conclusively. There will always be some kind of contradiction.
But I'm working on opening a separate issue about floating time, so we can focus here on the location implementation.

Caching location data is not as straightforward as it might seem, because the time zone database could get updated every now and then. You can imagine for example a server-process running for several years  and the tz database had 10 releases alone in 2016. As we all know, cache-invalidation is one of the most difficult things in computer science.

The best way to handle this depend highly on the application. I like the idea to make (optional) caching of location data a job of the application or a 3rd party library. The current solution works well enough for stdlib and should be sufficient for many simple use cases. And it's easy to build upon.

RX14 commented:
Floating doesn't have enough information to be compared, so it shouldn't. Always return false unless it's floating too.

larubujo commented:
this change makes all code (like an empty file or a hello world) require zip, zlib, flate, crc32. doesn't sound like a nice thing...

straight-shoota commented:
@larubujo Yes, thanks for noticing. This is only temporary to provide a prototype. These dependencies will be removed.
I'm planning on including a custom reader, like it is implemented in Go. This is pretty straightforward for uncompressed ZIP files, I just didn't want to spend the time on that when a working example can rely on the existing implementation.

straight-shoota commented on src/time/format/formatter.cr:
```diff
@@ -139,51 +139,55 @@ struct Time::Format
+
+      negative, hours, minutes, seconds = local_time_zone_info
+      io << (negative ? "-" : "+")
+      io << "0" if hours < 10
```
These lines were not modified by this PR (only whitespace change), so I think this could be fixed independently.

larubujo replied:
also char is slower than string. the people that thought that char is more efficient got it wrong. same when you index a string with a char instead of a string. or starts_with. char is slower. so string is preferred.

straight-shoota replied:
@larubujo Sure about this? My benchmark says otherwise. However, this is completely off-topic here. So, if you have some evidence to support your claim, please post a new issue about that. This would be quite unexpected.

Sija replied:
@larubujo it's true indeed for `starts_with?(Char)` overload, its performance was severely  degraded since *v0.23.1*:

```cr
Benchmark.ips do |x|
  x.report("starts_with? Char") { "foo".starts_with? 'f' }
  x.report("starts_with? String") { "foo".starts_with? "f" }
end
```

*v0.23.1*:
```
  starts_with? Char  457.3M (  2.19ns) (±19.03%)  1.02× slower
starts_with? String 468.03M (  2.14ns) (±17.47%)       fastest
```

*master*:
```
  starts_with? Char  207.7M (  4.81ns) (±13.71%)  2.23× slower
starts_with? String 463.22M (  2.16ns) (±13.89%)       fastest
```

Yet, passing a `Char` to `IO` is still faster:

```cr
Benchmark.ips do |x|
  io = IO::Memory.new

  x.report("io << Char") { io << 'f' }
  x.report("io << String") { io << "f" }
end
```

*v0.23.1*:
```
  io << Char 139.61M (  7.16ns) (±19.08%)       fastest
io << String  82.44M ( 12.13ns) (±20.64%)  1.69× slower
```

*master*:
```
  io << Char 142.37M (  7.02ns) (±17.83%)       fastest
io << String  83.37M ( 11.99ns) (±13.60%)  1.71× slower
```

larubujo replied:
it seems i was partially wrong. but it still slower for non-ascii:

```crystal
require "benchmark"

haystack = "がこう"

needle_string = "う"
needle_char = 'う'

a = 0
Benchmark.ips do |x|
  x.report("ends char") do
    a += haystack.ends_with?(needle_char) ? 1 : 0
  end
  x.report("ends string") do
    a += haystack.ends_with?(needle_string) ? 1 : 0
  end
end

Benchmark.ips do |x|
  x.report("starts char") do
    a += haystack.starts_with?(needle_char) ? 1 : 0
  end
  x.report("starts string") do
    a += haystack.ends_with?(needle_string) ? 1 : 0
  end
end

puts a
```

output

```
  ends char  95.17M ( 10.51ns) (±12.16%)  1.35× slower
ends string 128.26M (   7.8ns) (±17.20%)       fastest
  starts char  91.51M ( 10.93ns) (±11.07%)  1.45× slower
starts string 133.03M (  7.52ns) (±12.89%)       fastest
```

in any case times around nanosecond, so corrections like use '0' instead of "0" dont matter much

RX14 replied:
I agree that it doesn't matter much, but if you consider the future evolution of the standard library then using the char overload gives you much more optimization possibilities, even if the current performance is different. 

If you notice a performance problem, make an issue please.

ysbaddaden commented on src/time/location.cr:
```diff
@@ -0,0 +1,307 @@
+  end
+
+  # :nodoc:
+  def self.zone_sources
```
Just call `Crystal::System::Time.zone_sources` directly.

ysbaddaden commented on src/time/location.cr:
```diff
@@ -0,0 +1,307 @@
+  end
+
+  # :nodoc:
+  def self.load_localtime
```
Ditto: just call `Crystal::System::Time.load_localtime` directly.

straight-shoota commented on spec/std/time/location_spec.cr:
```diff
@@ -0,0 +1,265 @@
+      zone.dst?.should be_false
+    end
+
+    it ".local" do
```
I concur it is confusing, but so would it be if `TC=""` *did not* mean `UTC` in contrast to `tzset`... I'm not sure what would be better, but in doubt I'd rather mirror the behaviour of `tzset` instead of introducing new semantics.

straight-shoota replied:
I think the semantics of `TZ` environment variable should stay the same, we could perhaps remove `Location.load("")` to mean UTC.

straight-shoota commented on src/crystal/system/unix/time.cr:
```diff
@@ -62,6 +62,25 @@ module Crystal::System::Time
+  end
+
+  def self.load_localtime
+    if location = ::Time::Location.load("localtime", ["/etc/"])
```
I guess it could just open the file here and use `read_zoneinfo` directly, skipping a few unnecessary steps in between.

straight-shoota commented on src/time.cr:
```diff
@@ -181,12 +160,12 @@ struct Time
   end
 
-  def self.new(year, month, day, hour = 0, minute = 0, second = 0, *, nanosecond = 0, kind = Kind::Unspecified)
+  def self.new(year, month, day, hour = 0, minute = 0, second = 0, *, nanosecond = 0, location = Location::UNSPECIFIED)
```
Well, this wasn't true before - it just didn't matter much.

straight-shoota commented:
@ysbaddaden Thanks for the in-depth review!

1. Please see my previous comment answering to @larubujo. I'd implement the ZIP reader directly - it doesn't need to support compression, so it's quite easy to do (essentially a port of [this Go function](https://github.com/golang/go/blob/master/src/time/zoneinfo_read.go#L262)). No need to depend on zlib. On most unix systems, time zone data should be readily available. But on other platforms like Windows (which will be supported eventually) there will be a need for applications to have some other source for a tzinfo database, preferably a ZIP file.

2. Yes. This is surely open for debate. I've quickly touched this in a comment before:
> Caching location data is not as straightforward as it might seem, because the time zone database could get updated every now and then. You can imagine for example a server-process running for several years and the tz database had 10 releases alone in 2016. As we all know, cache-invalidation is one of the most difficult things in computer science.
I think the best strategy for stdlib is to to provide the basic tools to have proper time zone support, but leave optimization layers like caching and maybe some sort of location registry to the application or 3rd party libraries. That's also the way they do it in Go (though I'm not sure if there are libraries for caching locations - but the general idea that the stdlib implementation always reads from the database directly).
If you just need the current timestamp, you can prevent a lookup of the local timezone by using `Time.utc`.

3. We had `Kind::Unspecified` before, that's why I tried to incorporate this concept (even if it did not literally mean "floating time"). But I agree on your reasoning to remove it from this PR. Though I see a few (minor) issues that might need discussion. I will address that in a subsequent comment.

straight-shoota commented on spec/std/time/time_spec.cr:
```diff
@@ -138,6 +139,16 @@ describe Time do
+    with_zoneinfo do
+      location = Time::Location.load("Europe/Berlin")
+      reference = Time.new(2017, 10, 28, 13, 37, location: location)
+      next_day = Time.new(2017, 10, 28, 13, 37, location: location)
```
Yeah, that's a typo. This example is marked as `pending`, thus it's not executed at all (and the error got unnoticed until now). With the current implemenation of `Time::Span` it won't work, because `1.day` equals `24 hours`, but from `2017-10-28 13:37` (CEST +02:00) to `2017-10-29 13:37` (CET +01:00) it is only `23 hours` (but still `1 day`).

straight-shoota commented on src/time/location.cr:
```diff
@@ -0,0 +1,307 @@
+        return localtime
+      end
+    else
+      begin
```
I think it's sufficient to add `load?(name, sources)`. At least for this usage we don't need all the special case checks.

straight-shoota commented on src/time/location/loader.cr:
```diff
@@ -0,0 +1,134 @@
+  end
+
+  # :nodoc:
+  def self.load_from_dir_or_zip(name, source)
```
If I get a thumbs up, I'll happily implement the simple ZIP access.

straight-shoota commented:
Ad 3. Removing `Location::UNSPECIFIED`

It is fairly straightforward to remove all the parts relating to floating time. However, there is an issue with the time format parser: The parser can always encounter time format without time zone which are by themselves not related to a location. How this is to be interpreted depends on the context. In YAML for example, a time without offset or time zone name is considered to be in UTC. In other cases it might mean local time (though that is quite imprecise) or just be undefined (=floating time).
Now, even if we don't support floating time explicitly, it should be possible to recognize that a time value without time zone was parsed. I think this can be handled with user code, like creating a custom "unspecified" location object and using this as default value. If the parsed time value has this location, it did not contain a time zone.
The question is, should there be a default value for `Time::Format::Parser.time(location)` and if so, which would it be? I don't think we should *assume* any specific default, because this can't be determined generally. The culprit is that a default location is only needed for the relative uncommon parsing of a time format which does not include a time zone. For most use cases, it makes *no sense* to require a default location at all because it's simply never needed.
So, we could either select an arbitrary default location if none particular is specified by the caller, which could be completely wrong in some contexts but might go unnoticed, possibly introducing a subtle hidden bug.  Or we could remove the requirement to have an implicit default location which would raise an exception if none is specified by the caller. This would also fail but at least it would be noticed.
I'd prefer the latter: if the format does not include time zone information and no default location was explicitly provided, there is no way to know how such a value should be represented as an instance of `Time`.

RX14 commented:
Go for nillable default timezone with raise on get.

straight-shoota commented:
Next issue: This PR includes a change to the string representation of a `Time` instance to include the offset and location name (if available). Now that `Time.new` should return an instance in localtime, the result of `Time.new(2014, 10, 30, 21, 18, 13).to_s` changes from `2014-10-30 21:18:13` to `2014-10-30 21:18:13 +01:00 Local`. The actual offset and name obviously depends on local configuration. This is probably unexpected in many cases.

Possible solutions:
* A: Remove offset and location completely from all string representations
* B: Remove offset and location completely from `#to_s` but keep it for `#inspect`
* C: Remove offset and location if location is local
* D: Remove location name "Local" if location is local

I'd recommend C because it keeps close to the previous behaviour while adding the additional information provided by location if it differs from local. The only downside is that it is not immediately visible what time zone applies. This would speak for D, but it also makes the output less predictable and I believe for the common use of times created with `Location.local` it probably doesn't matter that much to directly include the time zone and offset.

asterite commented:
@straight-shoota What are we missing to proceed with this? This PR is **excellent** and what I had in mind to implement Time (well, it's more or less like it's represented in Go). I guess we'd like to avoid depending on `zip` so I can try to write the simple flat zip parser for this if it's needed (though it should be simple to do by looking at how the `Zip` type is implemented).

asterite commented:
@straight-shoota Hm, actually, it seems you already did that 😊

straight-shoota commented:
Thanks. I guess it's rather complete now, assuming there are no objections against me chosing option C from my last post (which didn't receive any feedback so it seems everyone is fine with it). And yes, a custom zip reader is already implemented, it was quite fun ;)

So from my point I'm just waiting for reviews and approval. I'll probably squash some commits to make it easier to follow the logic steps instead if the WIP history.

asterite commented:
I think I prefer option D. If you serialize a time it at least keeps the offset info.

RX14 commented on src/time.cr:
```diff
@@ -481,13 +466,18 @@ struct Time
+      to_s "%F %T UTC", io
     when local?
-      Format.new(" %:z").format(self, io)
+      to_s "%F %T", io
```
I don't like that the string representation of time depends on the local timezone of the computer. The string representation should be purely UTC or an offset. Omitting the offset when the offset happens to be the local timezone will probably lead to bugs.

straight-shoota replied:
The offset is not omitted if it matches the local timezone but only if this instance has `location == Location.local` But as @asterite also preferes option D) I will update that.

RX14 commented on src/time.cr:
```diff
@@ -549,13 +539,21 @@ struct Time
     epoch.to_f + nanosecond.to_f / 1e9
   end
 
+  def in(location : Location) : Time
```
docs?

RX14 replied:
Also perhaps return early if `location == self.location`.

RX14 commented on src/time.cr:
```diff
@@ -726,35 +724,24 @@ struct Time
-    seconds, nanoseconds = compute_seconds_and_nanoseconds
-    offset = compute_offset(seconds)
-    {seconds, nanoseconds, offset}
+  protected def self.compute_seconds_and_nanoseconds
```
Please remove this and replace it with direct references to `Crystal::System`.

RX14 commented on src/time/location.cr:
```diff
@@ -0,0 +1,281 @@
+  # `/usr/share/lib/zoneinfo/Foo/Bar` or `/usr/lib/locale/TZ/Foo/Bar` on unix-based operating systems.
+  # * If `TZ` is not set, the system's local timezone data (`/etc/localtime` on unix-based systems) will be used.
+  # * If no time zone data is found, `UTC` is returned.
+  def self.local
```
I feel like this should be cached. This method is very much on the fast path for a lot of operations.

RX14 commented:
Also I'd love (but not require) if this PR documented every public method on `Time`. I think we should really aim for 100% documentation these days...

straight-shoota commented on src/time.cr:
```diff
@@ -1,6 +1,10 @@
 require "crystal/system/time"
 
-# `Time` represents an instance in time. Here are some examples:
+# `Time` represents an instance in incremental time. Here are some examples:
```
As in "no floating time". Every `Time` instance points to a specific point in time.

straight-shoota replied:
Maybe this should be rephrased. Or just left as it was before...?

RX14 replied:
I think it needs to be rephrased

straight-shoota replied:
Well I took this definition issued by the W3C https://www.w3.org/International/articles/definitions-time/#floating_time
Does anyone have a better term?

asterite commented:
I'd also suggestion overriding `inspect` for `Time::Location` to show the location name. Right now it shows all its instance vars which is not helpful.

asterite commented on src/time/location.cr:
```diff
@@ -0,0 +1,281 @@
+  #
+  # Example:
+  # `ZONEFILE=/path/to/zoneinfo.zip crystal eval 'pp Location.load("Custom/Location")'`
+  def self.load(name : String)
```
I would make this return `nil` if the location is not found. Raising an exception is expensive.

straight-shoota replied:
But then you can't just simply do `Time.new(..., location: Location.load("Europe/Berlin"))`.

straight-shoota replied:
I think it should be provided that a location can always be loaded (as long as it is a valid name).

asterite replied:
Right. Well, we could then have `Location.load?` that returns `nil`.

I actually thought this would be needed when parsing a timezone, but now I think it has nothing to do. So for now we can leave it like that (we can always add `load?` later if needed).

straight-shoota replied:
There is also `.load?` (undocumented) which is fast without an exception, allthough it expects `zone_sources` as second argument. But this could have a default value of `Crystal::System::Time.zone_sources`.

asterite commented:
It would also be interesting to know how we'll parse locations with `Time.parse`. Right now we are missing the `%Z` flag to parse it, but we'll want to add it. It seems Go doesn't do load a location here: it just creates a new fake location with that name and an offset of zero. I personally don't like it, despite it being simple/efficient (but maybe bug prone). On the other hand, Ruby's `strptime` works very well here. I checked their source code and it seems they have internal tables that map names to offsets, though I don't know how would that work regarding daylight saving (if someone could test this it would be nice). Ideally we'd like to parse the timezone correctly. Mmm... I think I'm now confused, "ARS" is -03:00, but the timezone info is "America/Buenos_Aires"... how do we go from "ARS" to that or viceversa?

asterite commented:
Also worth checking out:

```crystal
require "benchmark"
require "time"

Benchmark.ips do |x|
  x.report("Time.now") do
    Time.now
  end
end
```

Old:

```
Time.now   1.89M (530.19ns) (± 3.84%)
```

New (!!):

```
Time.now  17.75k ( 56.33µs) (± 1.88%)  
```

In Ruby:

```ruby
require "benchmark/ips"
require "time"

Benchmark.ips do |x|
  x.report("Time.now") do
    Time.now
  end
end
```

Output:

```
            Time.now      1.727M (± 3.8%) i/s -      8.680M in   5.032395s
```

We can't let this be like that. `Time.now` is super common.

It seems Go caches the local timezone to avoid constructing it over and over (but it seems changing `ENV["TZ"]` in the middle of the program has no effect). In Ruby in `Time.new` I don't see any code to load the location (I think Ruby doesn't handle this at all), maybe that's why it's so fast.

Maybe we could cache the last known value of `ENV["TZ"]` and reload the local location if that changes.

straight-shoota commented:
@asterite I'm not really sure it would make sense to support anything else than timezone location names for parsing location data. I mean, that's what they're for and you wouldn't need them if it could be done with zones alone.

Zone abbreviations are essentially not much better than offsets because they don't identify a specific location and associated transitions.
For example, `WET` (Western European Time, UTC+00:00) could mean `Atlantic/Reykjavik`, `Europe/Lisbon` `Europe/London` or `Europe/Dublin` (maybe even some more). To make it even more complicated, Iceland doesn't observe daylight savings so they keep WET all year but the other locations change to `WEST` (Western European Summer Time, UTC+01:00).

asterite commented:
@straight-shoota So it's better to just do it like in Go then?

straight-shoota commented:
Yes. At least if `%Z` parses only a zone abbreviation, there is nothing you can do about it. Unless for some zones which might not have any ambiguity but I'm not sure that would be worth it. I'm not even to sure about the real use case of parsing location anyway. Most of the time this is just a presentation layer problem and exchanging time information should just use UTC (or a fixed offset timezone).

straight-shoota commented:
Chaching is really a difficult issue... It's certainly easy to cache the last location loaded from `ENV["TZ"]` assuming the name does not change. Then it's still a question of how to receive changes in the timezone database but that's a rather minor issue.
But the `TZ` environment variable is not used that commonly and most calls to `Location.local` will still result in querying `/etc/localtime`. If that file is changed, this should be propagated in a running Crystal program.
Maybe we could cache the location based on file mtime. But that will still mean accessing the file system. 

I fear accessing the local timezone can never be fast or must be inacurate - or rather unaware of changes to system settings.
Fast timestamps are provided by `Time.utc_now` (on my system 300x faster than `Time.now`) or using a cached location like this:
```crystal
local = Location.local

x.report("Time.now") { Time.now(location: local) }
```

Both are even faster than `Time.now` in master.

asterite commented:
I don't think changing /etc/localtime is common, so I wouldn't care about that case. In fact, I'm not sure I care about changes in ENV (Go clearly doesn't), but it seems something super easy to do. I just think that Time.now should be almost as efficient as before.

RX14 commented:
We should do what go does, and I assume go caches everything and `Location.local` never changes. There should be a method to invalidate the cache though.

straight-shoota commented:
If you're only staying at one place, `/etc/localtime` should be very much fixed, yes. But if you're travelling this is not so much true. There are even [tools](https://github.com/cdown/tzupdate) to automatically update the location according to your geoposition. And even if changes are not that common, I expect the new location to propagate to applications if changed in the system configuration.

I'm not convicend that we should sacrifice that for no real benefit. We can document that plain `Time.now` is rather slow and whoever needs high performance shouldn't have any downsides from using `Time.utc_now` (or `Time.now(location)` for custom caching). But the other way around it would be quite cumbersome to achieve accurate results reflecting the current state of system config. It would require extra development effort to invalidate the cache all the time (which will frequently just be forgotten).

RX14 commented:
> no real benefit

But opening a file is an absolutely *huge* slowdown. At the very least let's cache the value for 5 seconds.

asterite commented:
Not only that, it allocates memory. Previously it didn't.

I'd still like to cache it and check for modifications of the env var and file mtime. It should be simple and it would give the best user experience.

Sija commented on src/time/location/loader.cr:
```diff
@@ -0,0 +1,210 @@
+    file.seek -ZIP_TAIL_SIZE, IO::Seek::End
+
+    if file.read_bytes(Int32, IO::ByteFormat::LittleEndian) != END_OF_CENTRAL_DIRECTORY_HEADER_SIGNATURE
+      raise InvalidTZDataError.new("corrupt zip file")
```
`corrupt zip file` -> `Corrupted zip file`

straight-shoota commented:
@RX14 Yeah I mean no real benefit over at least checking if the file was modified and using a chached version. This also avoids countless allocations of actually the same location object. It's currently potentially a huge memory leak if you run thousands of `Location.local` and keep references through stored time instances. I'll rework that to introduce caching.

straight-shoota commented:
Hm, I've implemented a cache for `Crystal::System::Time.load_localtime`, now we only need one for `ENV["tz"]` to improve `Location.local`. But I'm considering adding a general mtime-based cache for any `Location.load`. I guess it should work and would be the most general solution.

straight-shoota commented:
Btw. the specs on circle ci are failing because it has `ENV["TZ"]` set :/ This is really difficult to spec...

RX14 commented:
It shouldn't check `mtime` on every single call. People care about time enough for it to be implemented in to VDSO to completely avoid syscall overhead. I'm not willing to accept more than a 5% performance decrease in `Time.new`.

straight-shoota commented:
Then `Time.new` can't use a default location that depends on system configuration which inevitably requires file system access. There's no way this can be anywhere comparable to the speed of the current implementation. If we want to accurately use the operating system's location setting it simply can't be as fast as without any configureable location data at all.

RX14 commented:
@straight-shoota that's the point. We don't really care about accuracy regarding timezone *by default*. We should allow applications to bust the cache manually if they *do* care about accuracy over performance, but the default should be performance: load `Location.local` once and once only.

straight-shoota commented:
All right, then we should just store `Location.local` in `Location::LOCAL` and use that as default value for `Time.new` (instead of a direct call). This way you don't need any complicated cache invalidation. It's just enough to call `Location.local` to get a *fresh* result.

straight-shoota commented:
Or, maybe better cache the value in a class var than a constant so it could be updated.

RX14 commented:
@straight-shoota I meant just have a `Location.clear_cache` which would set `Location.@@local = nil`, and then then next `Time.new` call would re-cache it again. To me that's simple and the application has control over the caching behaviour.

drhuffman12 commented:
In addition to [or via] `Location.clear_cache`, what about something like `Time.now` to use cached zone and `Time.now!` to clear cache and re-query the zone? [or `Time.new` vs `Time.new!`, etc?]

straight-shoota commented:
I think the easiest way would be for `Time.new` (and `Time#to_local` and `Time#local?`) to just implicitly use the cached location if one is available without checking validity. `Location.local` would always validate the timestamp and update the cache if needed. This way you don't need a dedicated cache-clearing method, calling `Location.local` is enough.

The most common (and for a naive user unexpected) calls to `Location.local` are the implicit ones in `Time.new` (and `Time.local?`, `Time.to_local` which are not very commonly used). If these are avoided by using a cached value if available, we're good. But that can perfectly be hidden from external view. No need to expose an additional variable or method.
This way, the public facing `Location.local` can always produce accurate results (it's not always reloading the file but checking if it is till valid). This is not as fast as taking the value from cache. But using `Location.local` directly should be quite rare and happen only if there is a specific reason for it. And it can be cached userside quite easily if that's necessary.

straight-shoota commented:
@drhuffman12 I find it much easier to use `Time.now(Location.local)` than `Time.now!` because it actually shows what's happening instead of a cryptic exclamation marke which could mean anything in this context.

asterite commented:
The cache shouldn't be reset by just calling Location.local or any of the other methods. It should be `Location.clear_local_cache` or something like that. The use cases of reloading the local location are super strange. You have to have a running program that's moving around the world, or someone fiddling for some reason with the env var or the files.

Please let's always cache it like in Go. No method should implicitly clear this cache. But if we need this, it should be explicit like I mention above.

RX14 commented:
I agree entirely with ary, you're making this way more complex than it needs to be. Just implement it how we've described we want it implemented.

straight-shoota commented:
`Location.local` won't clear the cache, just check if it's still valid. That's an order of magnitude faster than re-loading the location data.
Plus we'll also want a caching behaviour for `Location.load` so that it returns the same instance unless the file has changed.

straight-shoota commented:
@asterite I don't think he use case is strange at all. Many people regularly move between time zones and need to adjust their system settings acordingly. It might also just be that the timezone database was updated while the application is running.

asterite commented:
What's a current Crystal application that could be affected by this? As far as I know, most Crystal apps are webservers, and webservers don't travel.

ysbaddaden commented:
I agree, loaded locations should be transparently cached forever, always returning the same singleton instance, to avoid reading the same information over and over. Not just for `local` or the latest loaded location, but for all loaded locations. I may have a webserver that changes the timezone for all requests to the user's location for example (see the wonderful `Time.use_zone(name) {}` from Rails).

Also, I may have a long running process, and IANA zoneinfo files could be updated while it's running; we should reload them at some point. No need to check on every invocation or every few seconds, they're seldom updated after all, but we should. Checking every some minutes or even hours sounds acceptable to me. Pushing the invalidation to the user is a boring burden: how are we supposed to know we should invalidate the location cache?

straight-shoota commented:
I've reconsidered this and I agree that `Location.local` should not just change due to external influences. It is not only used by `Time.new` but also `Time#local?` for example. It would be unexpected if the result from `time.local?` would suddenly change during the runtime of a program. 
`Location.local` is now a class property and will be initialized on first call. It resembels the local time zone used by the Crystal program and is by default initialized from system settings. But it can also be configured by the application using `Location.local = custom_location` at any point. Reloading the system settings is just `Location.local = Location.load_local`. I think this is preferable over a ominous `Location.clear_local_cache` method since it shows exactly what's happening. This way also allows to directly query the system's localtime without side effects on a global variable and an application can easily define a custom location that should be treated as "local".

Besides that I have also included caching for `Location.load` based on file modification time. So it still requires a file system lookup. I don't think this should have any additional complexity. It is called `load` and it should be expected to be slow.
Applications who use many locations a lot should implement their own cached registry, this could perfectly be in a shard. I might even take a stab at that. But I don't think this should belong in stdlib. For reference: Go doesn't do any caching for locations loaded by `LoadLocation`. So my proposal is already quite an improvement: it is faster as it usually doesn't read data, just checks mtime. And it can be used all over the place without polluting memory with countless duplicates of the same location instances. It's still not super-fast like a customized, automatically updated registry could be, but it should be perfectly fine for most usecases.

luislavena commented on spec/std/time/location_spec.cr:
```diff
@@ -0,0 +1,323 @@
+  def __cached_zone=(zone)
+    @cached_zone = zone
+  end
+
```
A small nitpick here: can we close `Time::Location` here and avoid nesting the specs into it?

Thank you.

straight-shoota replied:
Some examples use protected methods like `transitions` to ensure that the spec covers all expected behaviour. I think it's easier to wrap the specs in the class namespace than providing additional accessors. At least I can't find any pressing argument against it.

If it was easily avoidable, I sure wouldn't do it that way.

luislavena replied:
Thank you for taking the time and provide the details behind it. While I would prefer to avoid testing internals, I believe those can be later improved, so 👍 with current implementation.

Cheers.

straight-shoota commented:
When checking the docs I encountered another issue regarding string representation of local time:
There are many examples like `Time.new(2016, 2, 15) # => 2016-02-15 00:00:00`. Now, the string is actually `2016-02-15 00:00:00 +01:00` in my current timezone and differs according to system settings.
What should such examples show when the exact results are diverse?

asterite commented:
We could either use a specific location (probably UTC) or just have a helper method in the specs to check for time.to_s that would ignore the timezone in the string.

straight-shoota commented:
Specs are another thing (I'm using UTC as well as ugly matchers like `t.to_s.should eq("2014-01-30 21:18:13 #{t.to_s("%:z")}")` [code](https://github.com/crystal-lang/crystal/pull/5324/files#diff-9cece23fd9a6c72b59845f1488681d6cR290) in the time specs) but I was talking about API docs for `Time` and others.

asterite commented:
API docs can use whatever timezone we want. We can say "assuming the local time is ..." before code examples.

waj commented:
I'm jumping quite late in the conversation, but let me add my two cents here with some questions:

1. If the `Time.location` is a pointer to the `Location` object, would that mean that if the locations are reloaded that instance of `Time` would still be pointing to an outdated location spec? Not big deal, but maybe it raises a design consideration here.
2. Location sounds to me like a precise coordinates to some place. For time zones the common noun used is, well... zone :o). Wouldn't `Zone` be a better name for the area where a current time applies?
3. I'm currently programming a lot with Elixir and there is an almost standard library to deal with timezones (https://github.com/lau/tzdata) that includes it's own tz database. It also periodically downloads the latest version so a running process can always have the latest specifications without having to wait until a distro maintainer updates the available packages (and the sysadmin install them). I'm not saying we have to ignore the system database, but I think it would be great to at least have the option to load the tz database from a custom location. Well... I didn't check all the source code and maybe this is already considered in the design.
4. How much time it takes to parse the tz database? If every `Time` instance now must point to a real location then every Crystal process must parse the tz database at least to get the local timezone? Considering this I'm all in favor of having an "unspecified" timezone, specially for those apps that don't care about timezones and maybe just uses times to render logs.
5. How is the performance now that the struct is bigger? The current `Time` struct fits into two 64bit registers so it works great to pass and return from functions. An option to avoid making the structure bigger is replace the pointer to the location with an index into a locations table. And maybe this could be useful to fix the issue I mention in point 1 by reloading the timezones into the same table in the exact same order. All this would require proper benchmarks and analysis of the generated code to figure out if the extra complexity is actually necessary.

waj commented:
Regarding point number 3, I just checked the sources and seems like it is already considered to load the data from a different location with an environment variable 👏

waj commented on src/time/location.cr:
```diff
@@ -0,0 +1,295 @@
+      # much less dot dot. Likewise, none begin with a slash.
+      raise InvalidLocationNameError.new(name)
+    else
+      if zoneinfo = ENV["ZONEINFO"]?
```
Shouldn't be `ZONEFILE` here?

straight-shoota replied:
I followed Go's lead here and I don't think `ZONEFILE` makes much sense because it could be either a folder or a (ZIP) file.

waj replied:
Oh, that's fine, but the docs should say `ZONEINFO` then 😄

straight-shoota replied:
Hehe, true

straight-shoota commented:
Hey @waj great to hear your voice on this issue!

1. Yes that's true. If the tz database is changed, `Location.load` will create a new instance and all `Time` instances still using the "old" value won't get updated. But I think this can be improved... If a timezone ID is already in the cache but outdated, the loader could probably just update the existing object instead of creating a new one. That requires some changes but can be done quite easily. Yet I'm not entirely shure it should work that way. Maybe we'd better leave this to a shard implementing a sophisitcated location cache/registry.
2. The terminology is quite confusing, really. "zone" can generaly refer to a region with standadized time or an actual time in effect like `BST`, `MET`, `ART`  etc. (or a fixed offset like `+03:00`). The latter is modeled in the class `Time::Location::Zone`.
"location" is the official name of the second part of a name in the timezone database. That's why I have chosen it as name of the type, following the terminology used in `Go`.
It is essentially a place on earth representing a region where at each point in time a specific time zone is in effect. Naming this `Zone` would be confusing because of ambiguity. We could also go for something else... I'm open. In  Java JSR-310 it's called `ZoneId`, in Rust chrono it's `TimeZone`... but I feel really strong about `Location` being a great name to express it's purpose.
3. Yes, it's configureable to use a custom path (either a dir or zip file).
4. This implementation doesn't parse the entire timezone database. Locations are loaded individually on demand. So, if you don't care about time zones, this will only mean a single location is loaded once (`Location.local`). Even this can be prevented by explicitly using predefined location (`Location::UTC`) or a custom created location (`Location.fixed`).
5. I don't have conclusive data on possible performance impact but to me it doesn't seem to be a real issue. I'd initially given the idea of a timezone index some thought but it didn't look very promising. I can't remember my reasoning exactly but I'm pretty sceptical about such a complex feature to *maybe* get a little performance improvement. I don't think I have seen any implementation doing this. Anyway, if we really find it needs to be improved, this can always be done later, it's just internal.

waj commented:
@straight-shoota all great responses! I'm totally ok with this. If something can be improved later... what can't? But it's an absolutely great starting point for a feature we're heavily missing.

RX14 commented on src/crystal/system/win32/time.cr:
```diff
@@ -11,14 +11,6 @@ module Crystal::System::Time
   NANOSECONDS_PER_SECOND    = 1_000_000_000
   FILETIME_TICKS_PER_SECOND = NANOSECONDS_PER_SECOND / NANOSECONDS_PER_FILETIME_TICK
 
-  # TODO: For now, this method returns the UTC offset currently in place, ignoring *seconds*.
```
Since time currently works on windows, can we not break it on windows? Having a working win32 implementation of this would make me confident that the `Crystal::System::Time` abstraction you've chosen works on windows.

straight-shoota replied:
Would it be ok with you if I assured you that I have a more-or-less working implementation for windows waiting? It's just missing conversion from `WCHAR*` to `String` for `Location.load_local` and `File` implementation for `Location.load`.
For obvious reasons, `Crystal::System::Time.zone_sources` will be empty on Windows because the OS doesn't include a copy of the tzdatabase.

RX14 replied:
so how will zone sources work on windows?

asterite replied:
At least in Go they seem to ship with a `zoneinfo.zip` in the installation, and check if that exists as a last resource: https://github.com/golang/go/blob/master/lib/time/zoneinfo.zip

luislavena replied:
AFAIK TimeZone information is available in Windows in the registry (`HKLM\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Time Zones`). Not sure how that can be made accessible to Crystal instead of `zoneinfo.zip` mechanism, or if that is relevant as mechanism to retrieve timezone information on Windows vs non-Windows platforms.

Cheers.

RX14 replied:
This should be done natively. Problem is that that registry data is probably a different format, which is why I bought up this issue.

straight-shoota replied:
Getting useful timezone data in Windows is a nasty problem. That's enough headeache to fuel another discussion. Basically: it's shit and you can't do anything about it.

Essentialy it comes down to two options: You either distribute a copy of the timezone database with your application or you can't use locations except UTC and local.

@luislavena You could try to get something out of Windows' time zone registry but that wont get you far. Windows just doesn't provide complete information. It contains only the standard and dst (if available) offset and dates applicable for the current year. That's not even worth trying, considering that the names are completely different from the IANA database. That would require a translation table to be able to use the same location names for windows and unix. Then it's much easier and better to just provide a copy of the tzdatabase. It's about 350 KB, so no big deal regarding distribution size. The main issue is that it's your responsibility to see it get updated.

@asterite That is not the last ressort but as far as I can tell the *only* way except manually providing a path to a database via `ZONEINFO` envvar. Otherwise it only works if you have Go installed (see go/golang#14276). - which will obviously not be the case on all machines running applications written in Go.

asterite replied:
So if I understand it correctly, there would be no good Time::Location support out of the box for Windows, and it's your responsibility to distribute that zip file with your application. I guess that's fine with me. We can improve this stuff later, but the only other option would be to somehow embed this data in the application, but that's probably way too much data.

RX14 replied:
Well, we should provide a way to optionally embed the zip data inside the executable on windows, but I'll accept your point that it's impossible.

straight-shoota commented:
I've added two commits to get this running on windows. 💯 Only `UTC`, `Location.local` and custom locations are supported for the moment. But loading tzdata files should work once `File` is properly ported.

Sija commented on src/crystal/system/win32/time.cr:
```diff
@@ -39,4 +42,119 @@ module Crystal::System::Time
+
+    time += (day - 1).days
+
+    return time.epoch
```
redundant `return`.

RX14 commented on scripts/generate_windows_zone_names.cr:
```diff
@@ -0,0 +1,76 @@
+
+nodes = xml.xpath_nodes("/supplementalData/windowsZones/mapTimezones/mapZone[@territory=001]")
+
+entries = [] of Tuple(String, Tuple(String, String), String)
```
Can we use named tuples here for descripriveness?

RX14 commented on src/lib_c/x86_64-windows-msvc/c/winbase.cr:
```diff
@@ -41,6 +41,15 @@ lib LibC
     daylightBias : LONG
   end
 
+  # Daylight saving time is not used in the current time zone, because there are no transition dates or automatic adjustment for daylight saving time is disabled.
```
We don't typically doc these...

RX14 commented on src/time/location/loader.cr:
```diff
@@ -23,6 +23,9 @@ class Time::Location
   # :nodoc:
   def self.load_from_dir_or_zip(name : String, source : String)
+    {% if flag?(:win32) %}
+    raise NotImplementedError.new("Time::Location.load_from_dir_or_zip")
```
Please indent these properly.

straight-shoota commented:
@RX14 fixed the issues.

straight-shoota commented:
Awesome to see this merged! Thanks for all the reviews and suggestions helping to get this done 💯
